### PR TITLE
Wire the tagged txmeta trim-ratio setting into the preallocated cache

### PR DESCRIPTION
**Pull request by icellan**
## Summary

Two direct `gocore.Config()` reads in `stores/txmetacache` bypassed the typed settings system, both fixed in `NewTxMetaCache`:

**1. Trim ratio.** The Preallocated bucket's trim ratio was controlled by an untagged `gocore.Config().GetInt("txMetaCacheTrimRatio", 2)` read in `improved_cache.go`, which bypassed the typed settings system entirely. The properly tagged `subtreevalidation_txMetaCacheTrimRatio` setting (documented, portal-visible, default `2`) had zero consumers.

`New` now takes an explicit, optional `trimRatio` parameter (variadic, matching the existing override pattern used elsewhere in this package), and `NewTxMetaCache` passes `tSettings.SubtreeValidation.TxMetaCacheTrimRatio` through. The tagged setting is now authoritative; the untagged key is gone and has no effect.

*Scoping note*: this knob only matters when `subtreevalidation_txMetaCacheBucketType` is `preallocated`. The default is `unallocated`, and nothing in `settings.conf`, `deploy/`, or `compose/` currently sets `preallocated`, so this path is inert in the current deployment. This PR fixes the wiring only; it does not attempt to determine whether `2` or `5` is the better trim ratio for the preallocated path — that requires a benchmark that hasn't been run, and can't be measured in situ while the path is inert.

`settings.conf`'s stray `txMetaCacheTrimRatio = 5` line is migrated to `subtreevalidation_txMetaCacheTrimRatio = 5` rather than deleted, so the previously configured value is preserved under its correct key instead of silently reverting to the code default of `2`. If `preallocated` is ever enabled, `5` is what will apply post-fix, not `2`.

The tagged field's longdesc claimed it "Passes through to bucketPreallocated.Init as trimRatio" — that was false before this change (zero consumers) and is now accurate.

**2. Cache size.** `NewTxMetaCache` also read the base cache size via `gocore.Config().GetInt("txMetaCacheMaxMB", 256)` instead of using the already-available `tSettings`. Unlike the trim-ratio setting, this tagged setting's key (`txMetaCacheMaxMB`, `settings/subtreevalidation_settings.go`) has **no** `subtreevalidation_` prefix — identical to the string the gocore call read, with the same default (256). Both paths resolved the same config value before this change, so this is a pure bypass removal with no behaviour change: `maxMB` is now read from `tSettings.SubtreeValidation.TxMetaCacheMaxMB`. The caller-supplied `maxMBOverride` still wins when provided, unchanged.

Worth a separate look: the prefix inconsistency between the two sibling settings in that struct (`txMetaCacheMaxMB` unprefixed vs `subtreevalidation_txMetaCacheTrimRatio` prefixed) is a naming smell, but renaming the key here would break anyone who has it set — not done in this PR.

This was the last use of the `gocore` import in `txmetacache.go` (and already removed from `improved_cache.go` by the trim-ratio commit); dropped.

## Test plan
- [x] `TestImprovedCache_New_TrimRatio`: explicit trimRatio honoured, omitted falls back to the default, non-preallocated bucket types ignore it
- [x] `TestBucketPreallocated_CleanLockedMap_TrimRatioAffectsRetainedEntries`: proves trimRatio actually changes eviction — a higher ratio retains fewer live entries after an identical write pattern
- [x] `TestNewTxMetaCache_TrimRatioFromTaggedSetting` / `TestNewTxMetaCache_UntaggedKeyHasNoEffect`: tagged setting reaches the cache via `settings.NewSettings()`; old untagged key has no effect
- [x] `TestNewTxMetaCache_MaxMBFromTaggedSetting`: tagged `txMetaCacheMaxMB` setting (via `settings.NewSettings()`, not a struct literal) drives the actual allocated cache size
- [x] `TestNewTxMetaCache_MaxMBOverrideWinsOverSetting`: `maxMBOverride` still takes precedence over the setting
- [x] `go build ./...`
- [x] `go vet ./stores/txmetacache/... ./services/subtreevalidation/... ./settings/...`
- [x] `go test -race -tags testtxmetacache ./stores/txmetacache/...`
- [x] `go test -tags testtxmetacache ./services/subtreevalidation/...`
- [x] Existing `BenchmarkImprovedCache_*` benchmarks run clean, no regression

### CODE REVIEW

**Review by ctnguyen — COMMENTED** [review]
Review at commit [eee67a9](https://github.com/icellan/teranode/commit/eee67a9b452d1415a3b58fcd1bca49ed21e34cac)

Good catch on the underlying bug — a tagged, documented, portal-visible setting with zero consumers while the value actually in use came from an untagged `gocore` read is exactly the kind of drift that survives for years. Migrating `settings.conf` instead of silently dropping to the default is the right call, the variadic follows the shape the package already uses, and three of the four new tests pin the wiring end to end. My concern is not with what the change does but with what it now exposes: this value is a percentage feeding unguarded chunk-rotation arithmetic, and the PR is what invites operators to touch it.

## Blocking issues

### ChiR1 — No bound on the newly authoritative trim ratio; out-of-range values panic on the first arena rotation

**Problem:** `trimRatio` reaches `bucketPreallocated.Set` with no validation at any layer — `settings/settings.go:649` is a plain `getInt`, `New` takes `trimRatio[0]` verbatim (`stores/txmetacache/improved_cache.go:320-323`), and `bucketPreallocated.Init` validates only `maxBytes` (`:1848-1874`). The value is consumed as a percentage at `:2010`: `numOfChunksToRemove = ceil(len(chunks) * trimRatio / 100)`, then `numOfChunksToKeep = len(chunks) - numOfChunksToRemove` at `:2011`. Anything outside `[1, 100]` indexes out of range on the first rotation, and the bound is independent of the chunk count:

  - `trimRatio <= 0` gives `numOfChunksToRemove <= 0`, so `numOfChunksToKeep >= len(chunks)`. At exactly `len(chunks)` both loops are no-ops, `idx` at `:2024` lands one chunk past the end, and `chunks[chunkIdx]` at `:2043` panics. For values negative enough that `ceil` actually goes below zero (roughly `trimRatio <= -100/len(chunks)`), the shift loop panics earlier, at `:2015`, reading `chunks[0+numOfChunksToRemove]`.
  - `trimRatio >= 101` makes `numOfChunksToKeep` negative; the shift loop is skipped and the clear loop at `:2019-2021` starts at a negative index.

Values `1` through `100` are all safe, because `math.Ceil` guarantees at least one chunk is removed for any positive ratio. Derived by reading `:2007-2043`; `go test` was not available to me, so I did not execute it.

**Why it matters:** This is pre-existing in the sense that the old `gocore` read was equally unvalidated — the PR does not introduce the hazard. What it changes is who is invited to touch the knob: an undocumented key that nobody knew existed becomes a portal-visible setting whose `usage` reads "Higher = more aggressive trimming" and whose longdesc says "Increase to free chunk memory sooner under tight RAM", with no ceiling stated anywhere. The failure mode is the bad kind: not a startup error but a panic inside `Set` on the cache write path, at whatever arbitrary point the arena first rotates, so a typo (`= 0`, `= 1000`) yields a node that starts cleanly and dies later under load. The codebase already treats operator-facing settings on this exact path defensively — `resolveTxMetaCacheBucketType` (`services/subtreevalidation/Server.go:976-992`) falls back with a warning "so a typo never silently changes the deployed allocator", and `Server.go:280-282` clamps `TxPolicyRejectedCacheMaxMB`. A bound here is the house convention, not an extra ask.

I considered arguing this non-blocking on the grounds that the Preallocated path is inert in every committed deployment (`TxMetaCacheBucketType` defaults to `unallocated`, and nothing under `settings.conf`, `deploy/` or `compose/` selects preallocated). That bounds today's blast radius, but it is an argument about the state the PR is trying to change: the point of wiring the setting is that someone will flip the bucket type and then tune this. A settings-wiring PR is the only natural home for the bound, and it is four lines in a function the PR already edits.

**Fix:** Reject out-of-range values where `New` unpacks the variadic, mirroring the existing `maxBytes` check at `:305-307`. Scoping it to the `Preallocated` branch keeps other bucket types untouched:

  ```go
  case Preallocated:
      if ratio < 1 || ratio > 100 {
          return nil, errors.NewServiceError("trimRatio must be between 1 and 100 (percent of the chunk arena freed per rotation); got %d", ratio)
      }

      c.trimRatio = ratio
  ```

Failing at construction surfaces the misconfiguration at startup instead of mid-traffic. If a hard failure is considered too strict for a tuning knob, clamping with a warning log is acceptable too — but silently accepting a value that panics is not.

## Non-blocking issues

### ChiR2 — Raising the trim ratio migrates the preallocated arena onto the Go heap, which is the opposite of what the longdesc recommends

**Problem:** `bucketPreallocated.Init` mmaps the whole bucket once at `:1858` (`MAP_ANON|MAP_PRIVATE`, `stores/txmetacache/mmap_name.go:69`) and slices it into `chunks` at `:1863-1867`. On rotation, the clear loop at `:2019-2021` replaces each trimmed slot with `make([]byte, ChunkSize)` — fresh Go heap. The shift loop only moves existing slice headers, nothing ever restores a chunk to its mmap view, and `git grep Munmap` scoped to `stores/txmetacache/` returns nothing, so the mapping is never released. After roughly `len(chunks)/numOfChunksToRemove` rotations every slot is heap-backed while the fully written arena stays mapped and unreferenced. `trimRatio` is precisely what sets `numOfChunksToRemove`, so raising it gets there sooner: on the production build (`BucketsCount = 8192`, `ChunkSize = 4096`) with the committed `txMetaCacheMaxMB = 1024`, each bucket holds 32 chunks, so `2` needs 32 rotations, the committed `5` needs 16, and `50` needs 2.

**Why it matters:** The longdesc this PR rewrites still tells operators to "Increase to free chunk memory sooner under tight RAM". Increasing it frees nothing — the arena is fixed at `Init` and never shrinks — it only accelerates the point at which the process holds the arena and an equivalent amount of Go heap at the same time. Someone following that advice under memory pressure gets the opposite of what the setting promises. The same longdesc also never says the value is a percentage; "ratio" plus a default of `2` reads naturally as one-half, and there is no range given, which is how a `0` or a `1000` gets into a config file in the first place (see ChiR1).

**Fix:** In this PR, correct the wording: state that the value is the percentage of a bucket chunk arena discarded per generation rollover, give the valid range as 1 to 100, and drop or rephrase the "free chunk memory sooner" recommendation, since higher values trade entry retention for allocation churn rather than for RSS. Separately, the allocation itself is worth a follow-up — note for whoever picks it up that a naive `clear(chunks[i])` is wrong: after the shift loop, slots `[numOfChunksToKeep, len(chunks))` still alias buffers that are now also referenced from lower indices, so clearing in place would wipe live data. Recycling requires saving the first `numOfChunksToRemove` headers before the shift and re-placing them, cleared, at the tail.

### ChiR3 — The new eviction test does not exercise the production path it claims to prove

**Problem:** `TestBucketPreallocated_CleanLockedMap_TrimRatioAffectsRetainedEntries` (`stores/txmetacache/improved_cache_test.go:1028-1069`) documents itself as proving "trimRatio actually changes eviction behaviour", but it never calls `Set`. It defines a local `numOfChunksToRemove` closure at `:1032-1034` that re-implements the `improved_cache.go:2010` formula, then feeds the result straight into `b.cleanLockedMap(...)` at `:1063-1064`. `b.trimRatio` is assigned by `Init` but nothing on the assertion path ever reads it.

**Why it matters:** What the test demonstrates is that `cleanLockedMap` retains fewer entries when given a larger starting offset — true, but that is arithmetic about `cleanLockedMap`, not evidence that `trimRatio` reaches it. A regression that hardcoded `2` back into `:2010` would leave this test green, which is exactly the class of bug this PR exists to fix. The other three new tests (`TestImprovedCache_New_TrimRatio`, `TestNewTxMetaCache_TrimRatioFromTaggedSetting`, `TestNewTxMetaCache_UntaggedKeyHasNoEffect`) do cover the wiring properly, so the gap is confined to this one.

**Fix:** Either drive it through the real path — `Set` entries until the arena rotates, then assert on `len(b.m)` or `b.trimCount` across two buckets initialised with different ratios — or soften the comment to say it checks `cleanLockedMap` offset behaviour, so the next reader does not over-trust it.

### ChiR4 — The same untagged-gocore pattern survives one line above the call this PR fixes

**Problem:** `NewTxMetaCache` reads the cache size through `gocore.Config().GetInt("txMetaCacheMaxMB", 256)` at `stores/txmetacache/txmetacache.go:205`, immediately above the `New` call this PR corrects at `:221` — even though the function already holds `tSettings`, and `settings/settings.go:647` loads the very same key into `tSettings.SubtreeValidation.TxMetaCacheMaxMB`.

**Why it matters:** It is the same defect class the PR is fixing, in the same function. Two concrete consequences: `settings.NewSettings(alternativeContext...)` resolves via `gocore.Config(alternativeContext...)` (`settings/helpers.go:37-41`) while this raw read always uses the default context, so a non-default settings context silently gets the wrong cache size; and any caller or test that sets the struct field directly is ignored.

**Fix:** Take it from the typed settings, keeping the existing override branch intact:

  ```go
  maxMB := tSettings.SubtreeValidation.TxMetaCacheMaxMB
  if len(maxMBOverride) > 0 && maxMBOverride[0] > 0 {
      maxMB = maxMBOverride[0]
  }
  ```

The key and default are identical, so this is behaviour-preserving for the default context. Optional here, but it closes the same hole in the same function while it is open.

## Recap

| ID | Description | Required | Criticality |
|---|---|---|---|
| ChiR1 | Unbounded trim ratio panics on rotation | YES | 75% |
| ChiR2 | Preallocated arena migrates to heap | | 45% |
| ChiR3 | Eviction test bypasses production path | | 35% |
| ChiR4 | Adjacent untagged gocore maxMB read | | 20% |

**Review by ordishs — APPROVED** [review]
Verified the wiring end to end. Both fields are real settings, not tag-only declarations — `settings/settings.go:647` and `settings/settings.go:649` call `getInt` for them. `getInt` (`settings/helpers.go:37`) is `gocore.Config(ctx...).GetInt`, so for the default context the `maxMB` path resolves exactly as before; the "no behaviour change" claim holds. The key migration is complete: a repo-wide grep finds no remaining use of the untagged `txMetaCacheTrimRatio` in `deploy/`, `compose/`, or `docs/`.

Ran:

```
CGO_ENABLED=0 SETTINGS_CONTEXT=test go test -tags testtxmetacache -count=1 ./stores/txmetacache/... ./settings/...
ok  github.com/bsv-blockchain/teranode/stores/txmetacache  2.957s
ok  github.com/bsv-blockchain/teranode/settings            0.324s
```

`./services/subtreevalidation/...` did not build on my machine (`gobdk` needs cgo, no Xcode licence here), so I did not verify that package.

One item worth fixing before merge.

---

### stores/txmetacache/improved_cache.go:319-323 — P1: unvalidated trim ratio panics the cache

**Issue**: `New` accepts any `trimRatio` without validation, and `NewTxMetaCache` (`stores/txmetacache/txmetacache.go:220`) passes the operator setting straight through. The rotation path at `improved_cache.go:2010` computes `numOfChunksToRemove := ceil(len(chunks) * trimRatio / 100)`. Values of 0, negative values, and values above 100 all drive the following chunk index arithmetic out of range.

**Failure**: demonstrated, not latent. I built a Preallocated cache at each ratio and wrote entries until the chunk arena wrapped:

```
trimRatio=0   PANICKED: runtime error: index out of range [512] with length 512
trimRatio=-5  PANICKED: runtime error: index out of range [-25]
trimRatio=200 PANICKED: runtime error: index out of range [-512]
```

Trace for 0: `numOfChunksToRemove` is 0, so `numOfChunksToKeep` equals `len(chunks)`, `idx` becomes `ChunkSize * len(chunks)`, `chunkIdx` becomes `len(chunks)`, and `chunks[chunkIdx]` at line 2044 is out of range. For 200 the keep count goes negative and the clear loop at line 2018 indexes `chunks[-512]`.

**Reachability**: an operator must set `subtreevalidation_txMetaCacheBucketType` to `preallocated` and give `subtreevalidation_txMetaCacheTrimRatio` a bad value. The default bucket type is `unallocated`, so no current deployment hits this. But the field's own longdesc says "Increase to free chunk memory sooner under tight RAM" and states no upper bound. An operator who follows that documentation and sets 200 crashes the node on the first arena rotation. There is no `recover()` anywhere in `stores/txmetacache`, and the Kafka txmeta consumer calls `Set` on the hot path.

The old `gocore` read carried the same exposure. This PR is the right place to close it, because this PR is the one that makes the documented knob authoritative.

**Recommendation**: clamp in `New`, next to the existing default selection:

```go
ratio := defaultTrimRatio
if len(trimRatio) > 0 && trimRatio[0] > 0 && trimRatio[0] <= 100 {
    ratio = trimRatio[0]
}
```

A warning log on rejection would make a bad value visible. Returning an error instead is also defensible — it fails the node at startup rather than at first rotation. Please add a test for each of 0, a negative value, and a value above 100.

---

Smaller items, non-blocking:

- `stores/txmetacache/improved_cache_test.go:1030-1071` — `TestBucketPreallocated_CleanLockedMap_TrimRatioAffectsRetainedEntries` re-implements `numOfChunksToRemove` in the test body and calls `cleanLockedMap` directly. It never runs `bucketPreallocated.Set`, so it never exercises the rotation branch. Change the divisor at `improved_cache.go:2010` from 100 to 10 and this test still passes. Driving two buckets through `Set` until each wraps would cover the whole path.
- `stores/txmetacache/txmetacache_maxmb_test.go:43-44` — the cleanup writes `""` rather than restoring `settings.conf`'s `txMetaCacheMaxMB = 1024` (line 1427). Every later `settings.NewSettings()` in the package then sees an unparseable empty value and falls back to 256. Same pattern at `txmetacache_trimratio_test.go:39` and `:62`. Capturing the prior value with `gocore.Config().Get(key)` and restoring it would avoid ordering-dependent surprises later.
- The `maxMB` "no behaviour change" claim is true for the default context only. `getInt` resolves through `gocore.Config(alternativeContext...)`, but the old call was a bare `gocore.Config()`. A `Settings` built with an alternative context — `NewSettings("dev."+cfg.DevName)` at `cmd/teranodedev/main.go:417` and five sibling call sites — now honours a context-scoped `txMetaCacheMaxMB` override that the old read ignored. Dev and test tooling only; no production daemon changes. Worth narrowing the wording in the description.
- `docs/references/settings/services/subtreevalidation_settings.md` lists `TxMetaCacheMaxMB` at line 21 but never lists `TxMetaCacheTrimRatio`. The corrected longdesc is therefore not published anywhere an operator reads. The omission predates this PR, but regenerating the reference fits its stated purpose.
- Migration: operators who set `txMetaCacheTrimRatio` in their own `settings_local.conf` or in a deployment overlay outside this repo lose that value silently. Nothing in the repo still uses the key, so a release note should cover it.

**Comment by icellan** [comment]
Added a test-harness fix as a separate commit (b1478bf69) on top of the trim-ratio/maxMB wiring.

**Bug**: `testPrunerObserver` in `daemon/test_daemon.go` buffered every `OnPruneComplete` callback into a capacity-10 channel, and `waitForPrune` just popped whatever was next in it — "return the next completion", not "return the completion for what I just did." A test that triggers several prune cycles (e.g. mining a handful of blocks) without draining an event after each one accumulates a backlog. A later `WaitForPruner` call can then be satisfied by a stale, already-superseded completion while the pruning it's actually waiting on happens afterwards.

This wasn't hypothetical: `TestPrunerParentFullySpentNotDeletedBeforeChildren` hit it directly. Completions for heights 1-11 queued unconsumed; the final `WaitForPruner` drained one of those, and the parent deletion it was asserting on only happened at height 12 — flaky pass/fail depending on which stale event got popped. Since every test that calls `WaitForPruner` shares this observer, the bug affects the whole harness, not just that one test.

**Fix**: `WaitForPruner` now takes an optional `minHeight`. When given, the wait discards any buffered completion below it and only returns once it has seen one at or above it. Callers that don't pass a height keep the exact old "pop whatever's next" behavior, so the 3 other call sites (`block_persister_test.go`, `catchup_test.go`, `multi_node_pruner_manual_sync_test.go`) are unaffected. Only the two calls in `pruner_parent_before_child_test.go` (the file that surfaced the bug) now pass the current chain height.

Making the wait height-aware also turned the capacity-10 channel from a latent problem into an actual one: with a height-aware wait now capable of draining a long backlog in a single call, the channel filled and started silently dropping completions (non-blocking send, `default` branch) — including the exact completion the wait needed, which then timed out. Reproduced this in ~3/5 runs before fixing it. Replaced the channel with an unbounded slice guarded by a mutex plus a small coalesced wake-up channel, so completions are never dropped.

Verified: `go build ./...`, `go vet ./daemon/...`, `go test -tags testtxmetacache ./stores/txmetacache/...`, and `TestPrunerParentFullySpentNotDeletedBeforeChildren` passing cleanly across 5 separate invocations (it previously failed 3/5 with a naive height check before the channel fix). Also ran all 5 other `WaitForPruner` callers individually to confirm none hang with the new default behavior.

Folded this into the txmetacache PR rather than opening a separate one since it was found and needed while working in this area, and it's a pure test-harness change with no production code touched.

File: test/e2e/daemon/ready/pruner_parent_before_child_test.go [hunk 1]
```diff
@@ -385,15 +386,16 @@ func TestPrunerParentFullySpentNotDeletedBeforeChildren(t *testing.T) {
...
-		_, err := node.UtxoStore.Get(node.Ctx, parentHash, fields.DeleteAtHeight)
-		return errors.Is(err, errors.ErrTxNotFound)
-	}, 10*time.Second, 100*time.Millisecond, "fully spent parent should be pruned")
+	// Wait for the pruner to complete a cycle covering at least the current height,
+	// so a stale completion queued from an earlier (already-superseded) prune cycle
+	// can't satisfy the wait before the parent's deletion has actually happened.
+	node.WaitForPruner(t, 10*time.Second, lastBlock.Height)
+
+	_, err = node.UtxoStore.Get(node.Ctx, parentHash)
+	require.Error(t, err)
```
[the remark this answers is not in the document]
Comment (icellan) [reply test/e2e/daemon/ready/pruner_parent_before_child_test.go]: You are right, and that weakening was mine — the `errors` import got dropped during a rebase and the assertion was loosened to match, which is backwards.

Restored to `require.ErrorIs(t, err, errors.ErrTxNotFound)` in `d7b00afc0`, with the import back (teranode/errors, not stdlib). Ran the test against a real Aerospike testcontainer to confirm it still passes on the specific error rather than any error: `go test -tags testtxmetacache -race -run TestPrunerParentNotDeletedBeforeChildren ./test/e2e/daemon/ready/` -> ok, 16.5s.

**Review by ctnguyen — APPROVED** [review]
Review at commit [d7b00af](https://github.com/icellan/teranode/commit/d7b00afc069aab22cf8afc1d805827ceb44cd234)

Thanks for taking ChiR4 on. `NewTxMetaCache` now reads `txMetaCacheMaxMB` from `tSettings`, the override branch is untouched, and the `gocore` import is gone. The pruner harness rework is a real improvement too. Replacing the capacity-10 channel with a mutex-guarded slice plus a coalesced one-slot `notify` means completions can no longer be dropped, and the wake-up can't be lost either. An event appended between the drain and the `select` leaves a token in `notify`, and a stale token only causes one extra empty drain. The deadline is created once, so the timeout covers the whole wait. Callers that pass no height still get plain FIFO. `NotifyPruneComplete` fires after phase 2 has finished deleting (`stores/utxo/aerospike/pruner/pruner_service.go:920`), so swapping the polling `require.Eventually` for a height-aware wait followed by a single `require.ErrorIs` is sound. The main thing still open is ChiR1: the trim ratio is still unbounded at this head, so a bad value can still panic on the write path, and that needs to land before merge. ChiR2 and ChiR3 are unchanged, and there are two small new notes on the added tests below.

## Status of previous points

ChiR1 is still open. `New` copies `trimRatio[0]` without any check (`stores/txmetacache/improved_cache.go:320-323`). `bucketPreallocated.Init` just stores the value (`:1870`), and `settings/settings.go:654` is a plain `getInt`. So `0`, negative values and anything above `100` still index out of range in `Set` at `:2010-2043`, as the previous round described. @ordishs reproduced the same panic independently in his P1. One correction to the previous round's fix: it said that returning an error from `New` would surface the problem at startup, but that is not true on the real call path. `services/subtreevalidation/Server.go:270-273` assigns the result of `NewTxMetaCache` straight to `u.utxoStore`, only logs the error, and keeps going. The server then runs with a nil `utxo.Store`, and the first `u.utxoStore.GetBlockHeight()` / `BatchDecorate` call (`check_block_subtrees.go:559`, `services/subtreevalidation/processTxMetaUsingStore.go:127`) panics instead. To get a real startup failure, that branch also has to `return nil, err`. Without that change, clamping to the default with a warning log (the variant @ordishs suggested) is the safer option. ChiR2's `Fix` asked for a longdesc rewrite, and the text at `settings/subtreevalidation_settings.go:23` still says "Increase to free chunk memory sooner under tight RAM" and still gives no range or unit. ChiR3's test and its "proves trimRatio actually changes eviction" comment (`stores/txmetacache/improved_cache_test.go:1020-1070`) are unchanged. ChiR4 was fixed exactly as suggested, and the behaviour is the same for the default context. No committed config sets a context-suffixed `txMetaCacheMaxMB`.

| ID | Previous point | Required | Status |
|----|----------------|----------|--------|
| ChiR1 | Unbounded trim ratio panics | YES | NOT FIXED |
| ChiR2 | Longdesc advice and range | | NOT FIXED |
| ChiR3 | Eviction test bypasses Set | | NOT FIXED |
| ChiR4 | Untagged gocore maxMB read | | FIXED |

## Non-blocking issues

### ChiR5 — The new maxMB tests only pass under the testtxmetacache build tag
**Problem:** `TestNewTxMetaCache_MaxMBFromTaggedSetting` and `TestNewTxMetaCache_MaxMBOverrideWinsOverSetting` (`stores/txmetacache/txmetacache_maxmb_test.go:40`, `:77`) use sizes of 1, 2 and 4 MB and expect `(maxMB*1024*1024/BucketsCount)/ChunkSize` chunks per bucket. The file header acknowledges that this holds only if there is "no floor bump" (`:18-19`). With the test tag (`BucketsCount = 8`, `improved_cache_const_testing.go:10`) there is no bump and the numbers work. In the default build (`BucketsCount = 8 * 1024`, `ChunkSize = 4096`, `improved_cache_const_large.go:10,18`), every one of these sizes gives `maxBucketBytes` of 128 to 512 bytes. That is below `ChunkSize`, so `New` raises it to `ChunkSize * 8` (`improved_cache.go:316-317`) and `Init` creates 8 chunks. The expected value evaluates to 0, so `require.Equal(t, wantChunks, ...)` at `:58` and `:96` would fail. I worked this out by reading the code and did not run it.
**Why it matters:** `make test` passes `-tags testtxmetacache` (`Makefile:180`), and the PR unit-test workflow runs `make test` (`.github/workflows/teranode_pr_tests.yaml:132`), so the unit job stays green. Other tracked invocations run without the tag, for example `.github/workflows/nightly_tests.yaml:199` and the e2e/chaos Makefile targets. Beyond those, the untagged `go test ./...` listed under Verification in `AGENTS.md` would report two failures in this package. The rest of the package is written to be tag-agnostic (`stores/txmetacache/improved_cache_test.go:192` spells that out), so this file is the one that breaks the pattern.
**Fix:** Choose sizes that stay above the floor in both builds (`maxMB >= 32` when `BucketsCount = 8192`), or have the test compute the floor the same way `New` does:

  ```go
  wantChunksFor := func(mb int) int {
      perBucket := mb * 1024 * 1024 / BucketsCount
      if perBucket < ChunkSize {
          perBucket = ChunkSize * 8
      }
      return perBucket / ChunkSize
  }
  ```

Picking sizes above the floor is the better option. With the floor-aware helper, 1 MB and 2 MB both map to 8 chunks in the default build, so the `require.NotEqual` "different sizes must differ" check at `:70` would fail. With 32, 64 and 128 MB, the tests keep their meaning in both builds.

### ChiR6 — Config cleanup in the new tests wipes committed values instead of restoring them
*Same as [@ordishs's point "the cleanup writes `""` rather than restoring" (15-09-2026)](https://github.com/bsv-blockchain/teranode/pull/1746#pullrequestreview-5211749895).*
**Problem:** The new tests change process-global `gocore` keys and then reset them with `Set(key, "")`. That happens at `stores/txmetacache/txmetacache_maxmb_test.go:43-44` and `:80-81` for `txMetaCacheMaxMB`, and at `stores/txmetacache/txmetacache_trimratio_test.go:38-39` for `subtreevalidation_txMetaCacheTrimRatio`. `settings.conf` commits non-default values for both keys: `txMetaCacheMaxMB = 1024` at `:1472` and `subtreevalidation_txMetaCacheTrimRatio = 5` at `:1474`. Once the key holds an empty value, `getInt` falls back to the code default (256 and 2). So every later `settings.NewSettings()` in the package gets different values from the committed ones. The `Set(key, "")` cleanup pattern appears in other tests too; for example, `git grep` finds it in ten `_test.go` files under `settings/` and `daemon/`. What is different here is that this cleanup replaces a committed non-default value. The untagged-key cleanup at `txmetacache_trimratio_test.go:62-63` is harmless because that key is no longer in `settings.conf`.
**Why it matters:** Tests in this package that build their settings from `settings.NewSettings()` run with a 256 MB cache and a trim ratio of 2 if they happen to run after these tests, and with 1024 MB and 5 otherwise. Nothing fails today, but it is a hidden ordering dependency in a package whose tests are about cache sizing.
**Fix:** Save the previous value and put it back in the cleanup:

  ```go
  prev, _ := gocore.Config().Get(key)
  gocore.Config().Set(key, "1")
  t.Cleanup(func() { gocore.Config().Set(key, prev) })
  ```

## Recap

| ID | Description | Required | Criticality |
|----|-------------|----------|-------------|
| ChiR5 | MaxMB tests need build tag | | 20% |
| ChiR6 | Test cleanup wipes committed config | | 12% |
